### PR TITLE
Jitter buffer overflows

### PR DESCRIPTION
**Pull request by jdelapla**
*Issue #, if available:*

*Description of changes:*
1. Added tracking of tail timestamp of the buffer
2. Added overflow state tracking for when tail or head wrap around the MAX_UINT16 or MAX_UINT32 for the respective bit counts
3. Added considerations for timestamp overflow & underflow

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/source/PeerConnection/JitterBuffer.c [hunk 1]
```diff
@@ -82,29 +87,350 @@
...
+        if(pJitterBuffer->headTimestamp > pRtpPacket->header.timestamp) {
+            timestampDifference = pJitterBuffer->headTimestamp - pRtpPacket->header.timestamp;
+        }
+        else {
+            timestampDifference = (MAX_RTP_TIMESTAMP - pRtpPacket->header.timestamp) + pJitterBuffer->headTimestamp;
+        }
+
+        //1 frame per second, and 1 packet per frame, the most charitable case we can consider
+        //TODO track most recent FPS to improve this metric
+        maxTimePassed = pJitterBuffer->clockRate * seqNoDifference;
```
[the remark this answers is not in the document]
Comment (jdelapla) [reply src/source/PeerConnection/JitterBuffer.c:109]: Good bot.

File: src/source/PeerConnection/JitterBuffer.c [hunk 2]
```diff
@@ -82,29 +87,355 @@
...
+            timestampDifference = (MAX_RTP_TIMESTAMP - pRtpPacket->header.timestamp) + pJitterBuffer->headTimestamp;
+        }
+
+        //1 frame per second, and 1 packet per frame, the most charitable case we can consider
+        //TODO track most recent FPS to improve this metric
+        if((MAX_RTP_TIMESTAMP / pJitterBuffer->clockRate) <= seqNoDifference) {
+            maxTimePassed = MAX_RTP_TIMESTAMP;
+        }
+        else {
+            maxTimePassed = pJitterBuffer->clockRate * seqNoDifference;
```
[the remark this answers is not in the document]
Comment (jdelapla) [reply src/source/PeerConnection/JitterBuffer.c:113]: Bad bot. The check above this prevents this possibility.

File: tst/JitterBufferFunctionalityTest.cpp [hunk 3]
```diff
@@ -1120,6 +1130,501 @@ TEST_F(JitterBufferFunctionalityTest, latePacketsOfAlreadyDroppedFrame)
 
     clearJitterBufferForTest();
 }
+
+TEST_F(JitterBufferFunctionalityTest, timestampOverflowTest)
+{
```
Comment (niyatim23) [inline tst/JitterBufferFunctionalityTest.cpp:1135]: Do we want to remove this / add a TODO here?

File: tst/JitterBufferFunctionalityTest.cpp [hunk 4]
```diff
@@ -1120,6 +1130,501 @@ TEST_F(JitterBufferFunctionalityTest, latePacketsOfAlreadyDroppedFrame)
...
+                ASSERT_TRUE(FALSE);
+        }
+        EXPECT_EQ(0, mDroppedFrameIndex);
+    }
+
+    clearJitterBufferForTest();
+}
+
+TEST_F(JitterBufferFunctionalityTest, LongRunningWithDroppedPacketsTest)
+{
```
Comment (niyatim23) [inline tst/JitterBufferFunctionalityTest.cpp:1625]: Do we want to remove this / add a TODO here?

**Review by niyatim23 — APPROVED** [review]

**Pull request merged**